### ISSUE TITLE
pubsub: support providers that don't do acks

### ISSUE THREAD

**Issue by ijt** [google/go-cloud#1368]
NATS and Redis pubsub don't have acking, and it doesn't make sense to try to shoehorn them in by making message acks a no-op. That would mean the application code would be misleading.

To support pubsub providers without acks, I propose to add the following new types:

`pubsub.FAFSubscription`, whose `Receive` method returns a `pubsub.FAFMessage`
`pubsub.FAFMessage` that is mostly like a `pubsub.Message` but has no `Ack` method or associated bookkeeping

and corresponding interfaces within `pubsub/driver`. FAF here stands for "Fire and Forget".

Drivers would either implement `driver.Subscription` or `driver.FAFSubscription`, but probably not both.

**Comment by derekcollison** [comment]
Not sure I am liking FAF or that a message changes due to its Ack methods. If the message is truly at most once than Ack would be a no-op.

**Comment by ijt** [comment]
Fair enough. Let's try your approach.

**Comment by derekcollison** [comment]
I have implemented the driver and necessary interfaces but things of course do not pass, so will look to work through those tomorrow. Thx.

**Comment by ijt** [comment]
Great. Glad to hear about this progress.

**Comment by derekcollison** [comment]
Maybe a message should have knowledge if it requires an ack. pubsub.Message.RequiresAck() bool.

**Comment by ijt** [comment]
I'm not sure how that would be an improvement over having just pubsub.Message.Ack() that does a no-op on some providers. I can see a downside though: having to check that before calling Ack all the time.

**Comment by derekcollison** [comment]
Possibly not, but if I was reading someone's code and just saw an Ack that seemed to return no error I would probably think it did something. That make sense? I would rather see something like..

```go
if msg.RequiresAck() {
    msg.Ack()
}
```

**Comment by ijt** [comment]
I don't think that works well from the perspective of app developers. They would have to scratch their heads about whether to include that logic every time they write code that receives a message.

The evidence that I see so far points towards needing a separate Go CDK subpackage to support at-most-once providers. I filed an issue for that: #1378. Closing this issue as redundant with that one.

**Comment by jba** [comment]
Robert and I just revisited the ideas here. We now think it might work to support ack-less providers with the same API, like so:
- The driver interface will have a `HasAck` method (all names open to bikeshedding).
- The concrete `Subscription` type will also have a `HasAck` method that just delegates to the driver. (Note: the method is on `Subscription`, not `Message`.)
- `Ack` will panic if called when `HasAck` is false. This is harsh, but we need to be harsh so nobody ever thinks they are using an at-least-once system when they aren't; the failure mode in that case is severe (messages that should be processed never are).

Unfortunately, as Robert pointed out to me, there is another distinction between these two kinds of systems that is harder to capture in the API. If a subscriber disconnects from a system like NATS and then reconnects, it will miss any messages that are sent while it's away. That won't happen for, say, Google PubSub (unless it disconnects for a long time). Let's call this _persistence_. I don't know if I should be as worried about persistence as I am about Ack. I also don't know if there are systems that support one of these features but not the other; for example, persistence with at-most-once delivery.

@vsaroopchand, does Azure Service Bus with Receive-and-Delete qualify as an at-most-once system with persistence?

> I don't think that works well from the perspective of app developers. They would have to scratch their heads about whether to include that logic every time they write code that receives a message.

That's a good point, but I don't think it will be a problem in practice. When you design a system, you have to decide up front whether you want at-most-once or at-least once, and you will code accordingly. I actually think code like
```
if msg.HasAck() {
    msg.Ack()
}
```
indicates a poor design, which is why I don't think `HasAck` should be on `Message`. I think the main use of `HasAck` should look something like:
```
func main() {
    sub := foopubsub.OpenSubscription(...)
    if !sub.HasAck() {
        log.Fatal("need Ack")
}
```

**Comment by ijt** [comment]
@jba, what you have there is better than checking before every ack, but as far as I can tell it does nothing to support transient topics or subscriptions. It sounds like making the existing API more complicated without the gain in functionality that we would have by having a specialized API for at-most-once providers.

Acklessness is only one aspect of these providers, and not a very important one relative to their greater flexibility.

**Comment by vangent** [comment]
> support transient topics or subscription

What do think we need to do to support those? IIUC, if we provide `OpenTopic` and `OpenSubscription` constructors (like we do for other drivers), they will work. `mempubsub` does this today in fact.

**Comment by jba** [comment]
@ijt, what's the "gain in functionality"? It seems like an at-most-once API would have less functionality.

**Comment by jba** [comment]
@ijt, maybe one thing you mean is the extra convenience from not having to name subscriptions. Looking again at your PR (#1380), you have
```
Subscribe(ctx context.Context, topic string, callback func(msg []byte)) error
```
in the concrete type. But I noticed that NATS has the ability to have separate "queues" for a topic, where each queue gets all the messages (see [`Subscription.Queue`](https://godoc.org/github.com/nats-io/go-nats#Subscription)). To support that, we would probably want to have subscriptions created outside the concrete type, as we do for `pubsub`. That makes the pubsub lite API start to look more like the pubsub API, with separate Publisher and Subscriber resources created by driver functions.

**Comment by derekcollison** [comment]
I think persistence is an implementation detail. Where I think you are focused is whether or not a system will redeliver a message or not under different circumstances, e.g. not connected, dropped the message, crashed, etc. In those cases the clients need a way to tell the system they are done, i.e. Ack(). Some systems with callback semantics do this for you once the callback completes. So many times you have a choice of manual or auto Acks. Lastly you have systems that topics/subjects are created on the fly and are totally dynamic (with appropriate auth of course), like NATS, or where you need to take separate actions to create these before publishing or subscribing. That is also a possible clash point in the API. I was heavily involved with the JMS specification at TIBCO, so I understand your challenges here.

Finally, I do believe communications are driving technologies to deliver multiple patterns in a single tech, from RPC (with distributed queueing) to Pub/Sub (Data/Event Streaming). Most people use all the patterns when designing systems with NATS, so having go-cloud/pubsub only be pubsub would probably not appeal to them IMO.

**Comment by ijt** [comment]
@vangent, the way we have `OpenTopic` and `OpenSubscription` is with non-portable funcs that are not part of the concrete `pubsub` API. To bring the functionality I'm thinking of into the existing pubsub API, I guess there would have to be new types like `TopicMaker` and `SubscriptionMaker` with corresponding driver interfaces. But then there would have to be some way to allow for how this isn't something we'd likely support on the at-least-once providers. The complication and confusion it would cause doesn't seem worth it. That's why I'm saying this functionality should go in a separate API.

**Comment by derekcollison** [comment]
When you are trying to put a common API around multiple products and feature sets the result usually does not give you or your users really what they want. Creating a Topic/Subject a priori is product specific, some do and some all dynamic creation.

**Comment by ijt** [comment]
@jba, it's not just a matter of convenience. Being able to transiently subscribe is functionality that we don't currently support in a portable way. What you're suggesting is something that would only support it in an un-portable way, if I'm reading this right. But why do that when it's possible to do it in a portable way?

**Comment by jba** [comment]
> Being able to transiently subscribe is functionality that we don't currently support in a portable way.
> But...it's possible to do it in a portable way...

I think you could ask that about any of our APIs. Why not provide a portable way to make a `blob.Bucket`, for example? Or for that matter, non-transient topics and subscriptions?

**Comment by ijt** [comment]
@zombiezen pointed out offline that creating a topic or subscription within an at-least-once provider allocates resources that stick around on the server until something deletes them. So to write code for these providers that dynamically creates topics and subscriptions, you'd want to add something to garbage collect them. Using defer for that wouldn't work well because your app could be unceremoniously terminated. So then we would need some sort of garbage collection service to manage this stuff. Maybe it would be fine if we decided that the Go CDK is all-in on kubernetes and there could be a Go CDK garbage collector that has to be running in your cluster. I don't think we're likely to do that though, so trying to emulate at-most-once dynamism with at-least-once providers would probably lead to resource leaks.

**Comment by jba** [comment]
Here's our current design:
- The portable API won't change at all.
- Drivers that don't support ack will accept a function from the user to be used as a replacement. That function should be a required argument to the subscription constructor (`OpenSubscription` or `CreateSubscription`). The constructor should panic if the argument is nil.

Advantages:
- It's impossible to accidentally use an at-most-once system in place of an at-least-once system (because the constructor argument is required and cannot be nil).
- The user can decide whether ack should do nothing, panic, or do something else. We don't force that decision on them. (There are good arguments for both doing nothing and panicking.)

Details:
- Should the function be invoked as the body of `SendAcks`? Or should it be communicated back to the concrete implementation, so that all the work of acking (batching acks and invoking `SendAcks` can be avoided?
- What is the signature of the function? Is `func()` adequate, or should it take a `*Message` and/or return `error`?

**Comment by gedw99** [comment]
Regarding garbage collection using an extra service and hence needing to depend on K8 is a slippery slope.

**Comment by vangent** [comment]
@jba 	is this done?

**Comment by jba** [comment]
Yes, fixed by #1467.

**Comment by vsaroopchand** [comment]
Not sure I fully agree with the proposal. Basically, you are proposing SDK users understand the different permutation of settings to toggle Ack/NoAck. Why not simply support OpenSubscription, OpenSubscriptonNoAck (name up for debate) and let the SDK author setup the driver accordingly?

**Comment by vangent** [comment]
@vsaroopchand Not sure I understand your question. Yes, we could make separate constructors, but why is that better than a single constructor with `Options`?

The main decision we made here was to support drivers that do auto-ack, but require the user to provide a function that is run when `Ack` is called -- so they can decide whether to panic, do nothing, log a warning, etc.

So, if we had a `OpenSubscriptionAutoAck` constructor, one of its arguments would be that function.

**Comment by jba** [comment]
I think @vsaroopchand's point is about ergonomics: it's confusing to have an `ackFunc` argument which you pass `nil` to in order to get Ack to work, and non-nil if you don't want Ack. I think the answer is careful documentation and examples.

**Comment by vsaroopchand** [comment]
Thanks for the feedback, I'll go with the recommended solution (ackFunc) & documentation.

**Comment by derekcollison** [comment]
At most-once-systems also differ in that they do not need an ack on a publish. Just FYI, so if you are exposing only the subscription side that might not be enough.

**Comment by jba** [comment]
That is a good point. Ideally, the NATS driver would wait for the write buffer to be flushed and return the resulting error. If not, we should carefully document that it's asynchronous, and there should be some way to get the error that eventually may result.

**Comment by jba** [comment]
Does the NATS server produce any sort of acknowledgement when it receives a published message?

**Comment by derekcollison** [comment]
The protocol supports it but none of the clients ask for verbose mode, does not change semantics of the system and slows things way down obviously.

**Comment by jba** [comment]
The problem is the same one as with ack: if someone thinks Send is guaranteed, and they start with a provider where it is, and then switch to one where it isn't, their system may break in subtle ways. Namely, messages that they think are safely in durable storage on the server are in fact not.

I think we need something like the ack function for topics. Maybe it could be an option that says whether you want guaranteed Send. You're not going to like this, but I think the default should be yes. @vangent thoughts?

**Comment by derekcollison** [comment]
Possibly. But I feel those truly familiar with at-least-once vs at-most-once do not need the additional complexity in the api.

**Comment by vangent** [comment]
My two cents:

* @jba is right that a user's system may break in subtle ways if they switch from an "at-least-once" to "at-most-once" provider.
* I'm not sure there's anything we can do about it other than documentation. There's no way to support "guaranteed send" on an "at-most-once" provider; e.g., even if we waited for positive acknowledgement from the server when using NATS, that's still no guarantee the message will be delivered.
* The ["replacement for ack" function](https://github.com/google/go-cloud/issues/1368#issuecomment-469931607) is useful because it forces users to be explicit about what they want to do when the application calls `Ack` (no-op, log warning, crash). I can't think of an analog for "send may fail without returning an error".

@jba, it's not clear to me how your "option for guaranteed send" works. For "at-least-once" providers, it's not needed, and for "at-most-once" providers, it can't ever be true.

I think we should add a section to the package doc for each driver with a header something like "Semantics" that says what kind of semantics the driver supports, and includes a pointer to the longer discussion on the portable type. Most drivers would just say "at least once" or "at most once", but some (like Azure Service Bus) would have a lengthier explanation because they support multiple semantics based on `Options`.

**Comment by derekcollison** [comment]
I agree things could change but that is present regardless of QoS. Any change in provider will potentially cause issues. APIs that try to cover many products always have this issue. 

Nit, but not important in the long run. I can do guaranteed delivery if I am sending request/reply semantics over an at-most-one system.

For the replacement ack, this just makes the value of the API a bit less, more provider specific code would need to be implemented when changing providers.

**Comment by derekcollison** [comment]
Let me know concrete next steps though to get this PR merged and off both of our plates ;)

**Comment by vangent** [comment]
I filed #1554 to add package doc about semantics. The NATS PR has been submitted. I think we're in agreement about not changing anything in the API for the semantics of `Send`.
